exec/java-exec/src/main/java/org/apache/drill/exec/record/VectorContainer.java

```diff
@@ -210,6 +214,109 @@ public void addCollection(Iterable<ValueVector> vectors) {
+      for (int vectorIndex = baseIndex; vectorIndex < wrappers.size(); vectorIndex++) {
+        ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+        ValueVector srcVector = probeSrcContainer.wrappers.get(vectorIndex).getValueVector();
+        destVector.copyEntry(recordCount, srcVector, probeSrcIndex);
+      }
+    }
+    recordCount++;
+    initialized = true;
+    return recordCount;
+  }
+
+  private int appendBuild(VectorContainer buildSrcContainer, int buildSrcIndex) {
+    // "- 1" to skip the last "hash values" added column
+    int lastIndex = buildSrcContainer.wrappers.size() - 1 ;
+    for (int vectorIndex = 0; vectorIndex < lastIndex; vectorIndex++) {
+      ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+      ValueVector srcVector = buildSrcContainer.wrappers.get(vectorIndex).getValueVector();
+      destVector.copyEntry(recordCount, srcVector, buildSrcIndex);
+    }
+    return lastIndex;
+  }
+  private void appendProbe(VectorContainer probeSrcContainer, int probeSrcIndex, int baseIndex) {
+      // int baseIndex = wrappers.size() - probeSrcContainer.wrappers.size();
+      for (int vectorIndex = baseIndex; vectorIndex < wrappers.size(); vectorIndex++) {
+        ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+        ValueVector srcVector = probeSrcContainer.wrappers.get(vectorIndex - baseIndex).getValueVector();
+        destVector.copyEntry(recordCount, srcVector, probeSrcIndex);
+      }
+  }
+  /**
+   *  A special version of appendRow for the HashJoin; uses a composite index for the build side
+   * @param buildSrcContainers The containers list for the right side
+   * @param compositeBuildSrcIndex Composite build index
+   * @param probeSrcContainer The single container for the left/outer side
+   * @param probeSrcIndex Index in the outer container
+   * @return Number of rows in this container (after the append)
+   */
+  public int appendRow(ArrayList<VectorContainer> buildSrcContainers, int compositeBuildSrcIndex, VectorContainer probeSrcContainer, int probeSrcIndex) {
+    int buildBatch = compositeBuildSrcIndex >>> 16;
+    int buildOffset = compositeBuildSrcIndex & 65535;
+    int baseInd = 0;
+    if ( buildSrcContainers != null ) { baseInd = appendBuild(buildSrcContainers.get(buildBatch), buildOffset); }
+    if ( probeSrcContainer != null ) { appendProbe(probeSrcContainer, probeSrcIndex, baseInd); }
+    recordCount++;
+    initialized = true;
+    return recordCount;
+  }
+
+  /**
+   * A customised version of the special appendRow for HashJoin - used for Left
+   * Outer Join when there is no build side match - hence need a base index in
+   * this container's wrappers from where to start appending
+   * @param probeSrcContainer
+   * @param probeSrcIndex
+   * @param baseInd - index of this container's wrapper to start at
+   * @return
+   */
+  public int appendOuterRow(VectorContainer probeSrcContainer, int probeSrcIndex, int baseInd) {
```

Comment:
Specific to HashJoin, move out of VectorContainer.

Reply:
Done ...